chore: charge for more l2 gas costs

just-mitch opened the pull request:
charge l2 gas for:
- avm startup
- note hashes in private
- nullifiers in private
- logs in private

fcarreiro commented on yarn-project/simulator/src/avm/avm_machine_state.ts:
```diff
@@ -46,6 +46,7 @@ export class AvmMachineState {
   constructor(gasLeftOrL2GasLeft: Gas | number, daGasLeft?: number) {
     if (typeof gasLeftOrL2GasLeft === 'object') {
       ({ l2Gas: this.l2GasLeft, daGas: this.daGasLeft } = gasLeftOrL2GasLeft);
+      this.l2GasLeft -= FIXED_AVM_STARTUP_L2_GAS;
```
Don't you want this for both branches?

just-mitch replied:
Yes, for sure. My mistake.

fcarreiro commented on yarn-project/simulator/src/avm/avm_machine_state.ts:
```diff
@@ -46,6 +46,7 @@ export class AvmMachineState {
   constructor(gasLeftOrL2GasLeft: Gas | number, daGasLeft?: number) {
     if (typeof gasLeftOrL2GasLeft === 'object') {
       ({ l2Gas: this.l2GasLeft, daGas: this.daGasLeft } = gasLeftOrL2GasLeft);
+      this.l2GasLeft -= FIXED_AVM_STARTUP_L2_GAS;
```
Are you aware making this change here in the machine state will also charge the cost for each external call (even each nested)? (even down the road when we do 1 enqueued call per AVM circuit)

just-mitch replied:
Yes, that was the intent, because as it stands we need to generate a public kernel proof for each external/nested call right? And that is what this startup cost is primarily intended to cover.

fcarreiro commented on yarn-project/simulator/src/avm/avm_machine_state.ts:
As I mentioned in the chat I think this would break (or at least misalign) the AVM circuit. (@dbanks12 correct me if I'm wrong).

Connecting to my other comment about doing this every enqueued call vs every nested call/avm circuit: did you consider making this gas consumption at the abstract phase manager or publicexecutor level?

Note that, on the one hand, that probably lets you just do the circuit equivalent in one of the public kernels. However, doing it in the abstract phase manager/public executor will not let you do this for every nested call (/current avm circuit execution).

just-mitch replied:
I wanted it to happen in the AVM so that the public kernel wouldn't need to start doing its own "out of gas" checks, since it presently just accepts the result of the AVM. I'm happy to discuss alternatives though!

dbanks12 replied:
For nested calls, I think this constant is unnecessary as it can just be considered part of the CALL & STATICCALL instruction cost.

This base cost might make sense for top-level enqueued calls since they AVM itself doesn't trigger a top-level call via an instruction.

Could that base gas instead be charged at the time a public call is enqueued in private land?

At the offsite we made substantial simplifications to how the AVM computes gas per-instruction because the circuit isn't ready for more complex gas computations. And we are now at the stage where we want the simulator and circuit to align. For now I believe we just a have a constant lookup table for gas (opcode -> const gas cost).

just-mitch replied:
@dbanks12  Ah yes of course!

Then it will follow the same pattern as notes/nullifiers where you get charged L2 gas in private, and we bake in additional L2 gas (32/64) when we EMITNOTEHASH/EMITNULLIFIER.

fcarreiro commented on yarn-project/simulator/src/avm/avm_machine_state.ts:
```diff
@@ -46,6 +46,7 @@ export class AvmMachineState {
   constructor(gasLeftOrL2GasLeft: Gas | number, daGasLeft?: number) {
     if (typeof gasLeftOrL2GasLeft === 'object') {
       ({ l2Gas: this.l2GasLeft, daGas: this.daGasLeft } = gasLeftOrL2GasLeft);
+      this.l2GasLeft -= FIXED_AVM_STARTUP_L2_GAS;
```
Can this potentially make the gas left negative or undeflow?

just-mitch replied:
Yes! Fixed this.

just-mitch commented:
Okay @dbanks12 @fcarreiro all AVM Simulator stuff is removed 👍 .

just-mitch committed:
charge l2 gas for:
- avm startup
- note hashes in private
- nullifiers in private
- logs in private

MirandaWood reviewed (approved):
Looks good for the protocol circuits!

just-mitch committed:
resolve conflicts
